connect/runtime/src/main/java/org/apache/kafka/connect/runtime/errors/RetryWithToleranceExecutor.java

```diff
@@ -0,0 +1,215 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.kafka.connect.runtime.errors;
+
+import org.apache.kafka.common.config.AbstractConfig;
+import org.apache.kafka.common.config.ConfigDef;
+import org.apache.kafka.common.config.ConfigException;
+import org.apache.kafka.common.utils.SystemTime;
+import org.apache.kafka.common.utils.Time;
+import org.apache.kafka.connect.errors.ConnectException;
+import org.apache.kafka.connect.errors.RetriableException;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.Locale;
+import java.util.Map;
+import java.util.concurrent.ThreadLocalRandom;
+
+import static org.apache.kafka.common.config.ConfigDef.Range.atLeast;
+import static org.apache.kafka.common.config.ConfigDef.ValidString.in;
+
+public class RetryWithToleranceExecutor implements OperationExecutor {
+
+    private static final Logger log = LoggerFactory.getLogger(RetryWithToleranceExecutor.class);
+
+    public static final String RETRY_TIMEOUT = "retry.timeout";
+    public static final String RETRY_TIMEOUT_DOC = "The maximum number of retries before failing an operation";
+    public static final long RETRY_TIMEOUT_DEFAULT = 0;
+
+    public static final String RETRY_DELAY_MAX_MS = "retry.delay.max.ms";
+    public static final String RETRY_DELAY_MAX_MS_DOC = "The maximum duration between two consecutive retries (in milliseconds).";
+    public static final long RETRY_DELAY_MAX_MS_DEFAULT = 60000;
+
+    public static final long RETRIES_DELAY_MIN_MS = 300;
+
+    public static final String TOLERANCE_LIMIT = "allowed.max";
+    public static final String TOLERANCE_LIMIT_DOC = "Fail the task if we exceed specified number of errors overall.";
+    public static final String TOLERANCE_LIMIT_DEFAULT = "none";
+
+    private long totalFailures = 0;
+    private final Time time;
+    private RetryWithToleranceExecutorConfig config;
+
+    public RetryWithToleranceExecutor() {
+        this(new SystemTime());
+    }
+
+    public RetryWithToleranceExecutor(Time time) {
+        this.time = time;
+    }
+
+    static ConfigDef getConfigDef() {
+        return new ConfigDef()
+                .define(RETRY_TIMEOUT, ConfigDef.Type.LONG, RETRY_TIMEOUT_DEFAULT, ConfigDef.Importance.HIGH, RETRY_TIMEOUT_DOC)
+                .define(RETRY_DELAY_MAX_MS, ConfigDef.Type.LONG, RETRY_DELAY_MAX_MS_DEFAULT, atLeast(1), ConfigDef.Importance.MEDIUM, RETRY_DELAY_MAX_MS_DOC)
+                .define(TOLERANCE_LIMIT, ConfigDef.Type.STRING, TOLERANCE_LIMIT_DEFAULT, in("none", "all"), ConfigDef.Importance.HIGH, TOLERANCE_LIMIT_DOC);
+    }
+
+    @Override
+    public <V> Result<V> execute(Operation<V> operation, ProcessingContext context) {
+        try {
+            switch (context.stage()) {
```

Comment:
could we condense this mapping in an enum or a static map such that we can avoid all this logic here?